src/services/codefixes/inferFromUsage.ts

```diff
@@ -665,29 +670,116 @@ namespace ts.codefix {
-                return checker.getUnionType([checker.getNumberType(), checker.getStringType()]);
+        const inferrers: ((k: UsageContext, checker: TypeChecker) => Type[])[] = [
+            inferFromPrimitive,
+            inferFromContext,
+            inferFromObject,
+            inferFromStructure,
+        ];
+
+        function getTypeFromUsageContext(k: UsageContext, checker: TypeChecker, fallback: Type): Type {
+            const inferences = [fallback];
+            for (const inferrer of inferrers) {
+                for (const t of inferrer(k, checker)) {
+                    inferences.push(t);
+                    updateInferenceProperties(t, k, checker);
+                }
             }
-            else if (usageContext.isNumber) {
-                return checker.getNumberType();
+            const anons = inferences.filter(i => checker.getObjectFlags(i) & ObjectFlags.Anonymous) as AnonymousType[];
+            const good = [];
+            if (!k.hasNonVacuousNonAnonymousType && anons.length) {
+                good.push(unifyAnonymousTypes(anons, checker));
             }
-            else if (usageContext.isString) {
-                return checker.getStringType();
+            good.push(...inferences.filter(i => !isVacuousInference(i, k, checker)));
+            return checker.getWidenedType(checker.getUnionType(good));
+        }
+
+        function unifyAnonymousTypes(anons: AnonymousType[], checker: TypeChecker) {
+            if (anons.length === 0) {
+                return checker.getNeverType();
+            }
+            if (anons.length === 1) {
+                return anons[0];
+            }
+            const calls = [];
+            const constructs = [];
+            const stringIndices = [];
+            const numberIndices = [];
+            let stringIndexReadonly = false;
+            let numberIndexReadonly = false;
+            const props = createMultiMap<Type>();
+            for (const anon of anons) {
+                for (const p of checker.getPropertiesOfType(anon)) {
+                    props.add(p.name, checker.getTypeOfSymbolAtLocation(p, p.valueDeclaration));
+                }
+                calls.push(...checker.getSignaturesOfType(anon, SignatureKind.Call));
+                constructs.push(...checker.getSignaturesOfType(anon, SignatureKind.Construct));
+                if (anon.stringIndexInfo) {
+                    stringIndices.push(anon.stringIndexInfo.type);
+                    stringIndexReadonly = stringIndexReadonly || anon.stringIndexInfo.isReadonly;
+                }
+                if (anon.numberIndexInfo) {
+                    numberIndices.push(anon.numberIndexInfo.type);
+                    numberIndexReadonly = numberIndexReadonly || anon.numberIndexInfo.isReadonly;
+                }
+            }
+            const members = mapEntries(props, (name, types) => {

```

Comment:
Could use some tests of combining different object inferences.

Reply:
I added one, and now I'm taking a look at the user tests for more examples.